File: backend-go/internal/service/plan_validation_service.go
package service

import (
	"github.com/leavemarker/backend-go/internal/repository"
	"github.com/leavemarker/backend-go/internal/utils"
)

// PlanValidationService handles plan limit validation
type PlanValidationService struct {
	employeeRepo     *repository.EmployeeRepository
	leavePolicyRepo  *repository.LeavePolicyRepository
	holidayRepo      *repository.HolidayRepository
	subscriptionRepo *repository.SubscriptionRepository
}

// NewPlanValidationService creates a new PlanValidationService
func NewPlanValidationService(
	employeeRepo *repository.EmployeeRepository,
	leavePolicyRepo *repository.LeavePolicyRepository,
	holidayRepo *repository.HolidayRepository,
	subscriptionRepo *repository.SubscriptionRepository,
) *PlanValidationService {
	return &PlanValidationService{
		employeeRepo:     employeeRepo,
		leavePolicyRepo:  leavePolicyRepo,
		holidayRepo:      holidayRepo,
		subscriptionRepo: subscriptionRepo,
	}
}

// ValidateEmployeeLimit checks if company can add more employees
func (s *PlanValidationService) ValidateEmployeeLimit(companyID uint) error {
	subscription, err := s.subscriptionRepo.FindActiveByCompanyID(companyID)
	if err != nil {
		return utils.NewBadRequestError("No active subscription found")
	}

	currentCount, err := s.employeeRepo.CountActiveByCompanyID(companyID)
	if err != nil {
		return utils.NewInternalServerError("Failed to count employees")
	}

	if int(currentCount) >= subscription.Plan.MaxEmployees {
		return utils.NewBadRequestError("Employee limit reached for your plan")
	}

	return nil
}

// ValidateLeavePolicyLimit checks if company can add more leave policies
func (s *PlanValidationService) ValidateLeavePolicyLimit(companyID uint) error {
	subscription, err := s.subscriptionRepo.FindActiveByCompanyID(companyID)
	if err != nil {
		return utils.NewBadRequestError("No active subscription found")
	}

	if subscription.Plan.MultipleLeavePolicies {
		return nil // Unlimited policies
	}

	currentCount, err := s.leavePolicyRepo.CountActiveByCompanyID(companyID)
	if err != nil {
		return utils.NewInternalServerError("Failed to count leave policies")
	}

	if int(currentCount) >= subscription.Plan.MaxLeavePolicies {
		return utils.NewBadRequestError("Leave policy limit reached for your plan")
	}

	return nil
}

// ValidateHolidayLimit checks if company can add more holidays
func (s *PlanValidationService) ValidateHolidayLimit(companyID uint) error {
	subscription, err := s.subscriptionRepo.FindActiveByCompanyID(companyID)
	if err != nil {
		return utils.NewBadRequestError("No active subscription found")
	}

	if subscription.Plan.UnlimitedHolidays {
		return nil // Unlimited holidays
	}

	currentCount, err := s.holidayRepo.CountActiveByCompanyID(companyID)
	if err != nil {
		return utils.NewInternalServerError("Failed to count holidays")
	}

	if int(currentCount) >= subscription.Plan.MaxHolidays {
		return utils.NewBadRequestError("Holiday limit reached for your plan")
	}

	return nil
}

// ValidateAttendanceManagementAccess checks if company has access to attendance management
func (s *PlanValidationService) ValidateAttendanceManagementAccess(companyID uint) error {
	subscription, err := s.subscriptionRepo.FindActiveByCompanyID(companyID)
	if err != nil {
		return utils.NewBadRequestError("No active subscription found")
	}

	if !subscription.Plan.AttendanceManagement {
		return utils.NewForbiddenError("Attendance management is not available in your plan")
	}

	return nil
}

// ValidateReportsDownloadAccess checks if company has access to reports download
func (s *PlanValidationService) ValidateReportsDownloadAccess(companyID uint) error {
	subscription, err := s.subscriptionRepo.FindActiveByCompanyID(companyID)
	if err != nil {
		return utils.NewBadRequestError("No active subscription found")
	}

	if !subscription.Plan.ReportsDownload && !subscription.HasReportDownloadAddon {
		return utils.NewForbiddenError("Reports download is not available in your plan")
	}

	return nil
}

// ValidateAttendanceRateAnalyticsAccess checks if company has access to attendance rate analytics
func (s *PlanValidationService) ValidateAttendanceRateAnalyticsAccess(companyID uint) error {
	subscription, err := s.subscriptionRepo.FindActiveByCompanyID(companyID)
	if err != nil {
		return utils.NewBadRequestError("No active subscription found")
	}

	if !subscription.Plan.AttendanceRateAnalytics {
		return utils.NewForbiddenError("Attendance rate analytics is not available in your plan")
	}

	return nil
}

// GetSubscriptionFeatures returns the subscription features for a company
func (s *PlanValidationService) GetSubscriptionFeatures(companyID uint) (map[string]interface{}, error) {
	subscription, err := s.subscriptionRepo.FindActiveByCompanyID(companyID)
	if err != nil {
		return nil, utils.NewBadRequestError("No active subscription found")
	}

	features := map[string]interface{}{
		"maxEmployees":            subscription.Plan.MaxEmployees,
		"maxLeavePolicies":        subscription.Plan.MaxLeavePolicies,
		"maxHolidays":             subscription.Plan.MaxHolidays,
		"attendanceManagement":    subscription.Plan.AttendanceManagement,
		"reportsDownload":         subscription.Plan.ReportsDownload || subscription.HasReportDownloadAddon,
		"multipleLeavePolicies":   subscription.Plan.MultipleLeavePolicies,
		"unlimitedHolidays":       subscription.Plan.UnlimitedHolidays,
		"attendanceRateAnalytics": subscription.Plan.AttendanceRateAnalytics,
		"advancedReports":         subscription.Plan.AdvancedReports,
		"customLeaveTypes":        subscription.Plan.CustomLeaveTypes,
		"apiAccess":               subscription.Plan.APIAccess,
		"prioritySupport":         subscription.Plan.PrioritySupport,
	}

	return features, nil
}

File: backend/src/main/java/com/leavemarker/service/PlanValidationService.java
package com.leavemarker.service;

import com.leavemarker.entity.Plan;
import com.leavemarker.entity.Subscription;
import com.leavemarker.enums.SubscriptionStatus;
import com.leavemarker.exception.BadRequestException;
import com.leavemarker.repository.EmployeeRepository;
import com.leavemarker.repository.HolidayRepository;
import com.leavemarker.repository.LeavePolicyRepository;
import com.leavemarker.repository.SubscriptionRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class PlanValidationService {

    private final SubscriptionRepository subscriptionRepository;
    private final EmployeeRepository employeeRepository;
    private final LeavePolicyRepository leavePolicyRepository;
    private final HolidayRepository holidayRepository;

    /**
     * Validates if the company can add more employees based on their plan
     */
    public void validateEmployeeLimit(Long companyId) {
        Subscription subscription = getActiveSubscription(companyId);
        Plan plan = subscription.getPlan();

        long currentEmployeeCount = employeeRepository.countByCompanyIdAndDeletedFalse(companyId);

        // For FREE plan, max 10 employees
        if (plan.getMaxEmployees() > 0 && currentEmployeeCount >= plan.getMaxEmployees()) {
            throw new BadRequestException(
                    String.format("Employee limit reached. Your current plan allows up to %d employees. Please upgrade to add more employees.",
                            plan.getMaxEmployees()));
        }
    }

    /**
     * Validates if the company can add more leave policies based on their plan
     */
    public void validateLeavePolicyLimit(Long companyId) {
        Subscription subscription = getActiveSubscription(companyId);
        Plan plan = subscription.getPlan();

        // For FREE plan, only 1 leave policy allowed
        if (!plan.getMultipleLeavePolicies()) {
            long activePolicyCount = leavePolicyRepository.countByCompanyIdAndActiveAndDeletedFalse(companyId, true);

            if (activePolicyCount >= plan.getMaxLeavePolicies()) {
                throw new BadRequestException(
                        String.format("Leave policy limit reached. Your current plan allows only %d active leave policy. Please upgrade to Mid Tier plan for multiple leave policies.",
                                plan.getMaxLeavePolicies()));
            }
        }
        // MID_TIER allows unlimited policies (no check needed)
    }

    /**
     * Validates if the company can add more holidays based on their plan
     */
    public void validateHolidayLimit(Long companyId) {
        Subscription subscription = getActiveSubscription(companyId);
        Plan plan = subscription.getPlan();

        // For FREE plan, max 6 holidays
        if (!plan.getUnlimitedHolidays()) {
            long holidayCount = holidayRepository.countByCompanyIdAndDeletedFalse(companyId);

            if (holidayCount >= plan.getMaxHolidays()) {
                throw new BadRequestException(
                        String.format("Holiday limit reached. Your current plan allows up to %d holidays. Please upgrade to Mid Tier plan for unlimited holidays.",
                                plan.getMaxHolidays()));
            }
        }
        // MID_TIER allows unlimited holidays (no check needed)
    }

    /**
     * Validates if the company has access to attendance management
     */
    public void validateAttendanceManagementAccess(Long companyId) {
        Subscription subscription = getActiveSubscription(companyId);
        Plan plan = subscription.getPlan();

        if (!plan.getAttendanceManagement()) {
            throw new BadRequestException(
                    "Attendance management is not available in your current plan. Please upgrade to Mid Tier plan to access this feature.");
        }
    }

    /**
     * Validates if the company has access to reports download
     */
    public void validateReportsDownloadAccess(Long companyId) {
        Subscription subscription = getActiveSubscription(companyId);
        Plan plan = subscription.getPlan();

        if (!plan.getReportsDownload() && !subscription.getHasReportDownloadAddon()) {
            throw new BadRequestException(
                    "Report downloads are not available in your current plan. Please upgrade to Mid Tier plan or purchase the report download add-on.");
        }
    }

    /**
     * Validates if the company has access to attendance rate analytics
     */
    public void validateAttendanceRateAnalyticsAccess(Long companyId) {
        Subscription subscription = getActiveSubscription(companyId);
        Plan plan = subscription.getPlan();

        if (!plan.getAttendanceRateAnalytics()) {
            throw new BadRequestException(
                    "Attendance rate analytics is not available in your current plan. Please upgrade to Mid Tier plan to access this feature.");
        }
    }

    /**
     * Gets the active subscription for a company
     * Throws exception if no active subscription found
     */
    private Subscription getActiveSubscription(Long companyId) {
        return subscriptionRepository.findByCompanyIdAndStatus(companyId, SubscriptionStatus.ACTIVE)
                .orElseThrow(() -> new BadRequestException("No active subscription found. Please subscribe to a plan first."));
    }
}
